File: packages/js/src/catalogue.ts
/**
 * The ngano catalogue: 612 African-language speech dataset records, bundled
 * with the package so that nothing here needs the network.
 *
 * @packageDocumentation
 */
import { CATALOGUE, COUNTRIES, LANGUAGES } from "./data/snapshot.js";
import { NganoHttpError } from "./errors.js";
import { compileFilter, fold, toOptions, type FilterLike } from "./filter.js";
import {
  languageName,
  primaryCode,
  resolveLanguage as resolveLanguageTags,
  type ResolveOptions,
} from "./languages.js";
import type {
  CountrySummary,
  Dataset,
  FetchLike,
  Language,
  LanguageSummary,
  RawCountry,
  RawDataset,
  Stats,
} from "./types.js";

/** Default base URL of the ngano HTTP API. */
export const API_BASE = "https://ngano.dev/api/v1";

/**
 * Converts a raw catalogue record into the camelCase public shape.
 *
 * @param raw a record from `catalogue.json` or from the HTTP API
 * @returns the public dataset object
 */
export function normaliseDataset(raw: RawDataset | Record<string, unknown>): Dataset {
  const record = raw as Record<string, unknown>;
  const str = (key: string): string | null => {
    const value = record[key];
    return typeof value === "string" && value.length > 0 ? value : null;
  };
  const list = (key: string): string[] => {
    const value = record[key];
    return Array.isArray(value) ? value.filter((v): v is string => typeof v === "string") : [];
  };
  const num = (key: string): number | null => {
    const value = record[key];
    if (typeof value === "number" && Number.isFinite(value)) return value;
    if (typeof value === "string" && value.trim() !== "" && Number.isFinite(Number(value))) {
      return Number(value);
    }
    return null;
  };
  const speakers = record["speakers"];

  return {
    id: str("id") ?? "",
    name: str("name") ?? "",
    task: str("task") ?? "Other",
    variety: str("variety"),
    languages: list("languages"),
    languagesClean: list("languages_clean").length ? list("languages_clean") : list("languages"),
    languageTags: list("language_tags"),
    languageCodes: list("language_codes"),
    iso: list("iso"),
    countries: list("countries"),
    countryCodes: list("country_codes"),
    regions: list("regions"),
    hours: str("hours"),
    hoursNum: num("hours_num"),
    speakers: typeof speakers === "number" || typeof speakers === "string" ? speakers : null,
    recordingType: str("recording_type"),
    quality: str("quality"),
    labelled: str("labelled"),
    domain: str("domain"),
    licence: str("licence"),
    licenceClass: str("licence_class"),
    commercial: str("commercial"),
    access: str("access"),
    host: str("host"),
    url: str("url"),
    hfRepo: str("hf_repo"),
    year: str("year"),
    notes: str("notes"),
    unverifiedSize: record["unverified_size"] === true,
    ...(str("language_note") !== null ? { languageNote: str("language_note") as string } : {}),
  };
}

/**
 * Hours that count towards a total. Self-reported figures of 20,000 hours or
 * more are excluded from every total, everywhere, per the ngano data policy.
 *
 * @param dataset a dataset
 * @returns hours to add to a total, zero when the figure is excluded
 */
export function countableHours(dataset: Dataset): number {
  if (dataset.unverifiedSize) return 0;
  return dataset.hoursNum ?? 0;
}

/** Options for {@link Catalogue.fromApi}. */
export interface FromApiOptions {
  /** API base URL. Defaults to `https://ngano.dev/api/v1`. */
  baseUrl?: string;
  /** `fetch` implementation. Defaults to the global one. */
  fetch?: FetchLike;
  /** Aborts the download. */
  signal?: AbortSignal;
  /** Page size, capped by the API at 200. */
  perPage?: number;
}

/**
 * A queryable view over ngano dataset records.
 *
 * @example
 * ```ts
 * const cat = new Catalogue();
 * cat.datasets({ language: "Shona", commercial: true, task: "ASR" });
 * cat.get("waxal-corpus-paper");
 * ```
 */
export class Catalogue {
  /** Every record in this catalogue, in source order. */
  readonly records: readonly Dataset[];
  private readonly byId: Map<string, Dataset>;
  private readonly byRepo: Map<string, Dataset>;
  private readonly countryIndex: Map<string, RawCountry>;
  private readonly languageIndex: Map<string, Language>;

  /**
   * @param records dataset records. Defaults to the bundled snapshot, which
   *   needs no network at all.
   */
  constructor(records?: ReadonlyArray<Dataset | RawDataset>) {
    const source = records ?? (CATALOGUE as RawDataset[]);
    this.records = source.map((record) =>
      "hfRepo" in record ? (record as Dataset) : normaliseDataset(record as RawDataset),
    );
    this.byId = new Map(this.records.map((record) => [record.id, record]));
    this.byRepo = new Map();
    for (const record of this.records) {
      if (record.hfRepo && !this.byRepo.has(record.hfRepo)) {
        this.byRepo.set(record.hfRepo.toLowerCase(), record);
      }
    }
    this.countryIndex = new Map();
    for (const country of COUNTRIES as RawCountry[]) {
      this.countryIndex.set(fold(country.name), country);
      this.countryIndex.set(country.iso2.toLowerCase(), country);
      this.countryIndex.set(country.iso3.toLowerCase(), country);
      if (country.map_name) this.countryIndex.set(fold(country.map_name), country);
    }
    this.languageIndex = new Map(
      (LANGUAGES as Language[]).map((language) => [language.tag.toLowerCase(), language]),
    );
  }

  /**
   * Downloads the live catalogue from the ngano HTTP API, paging until it has
   * every record. Falls back to nothing: if the API is unreachable the promise
   * rejects, and callers who want an offline catalogue should use
   * `new Catalogue()`.
   *
   * @param options API base URL, fetch implementation and abort signal
   * @returns a catalogue holding the live records
   */
  static async fromApi(options: FromApiOptions = {}): Promise<Catalogue> {
    const globalFetch = (globalThis as { fetch?: FetchLike }).fetch;
    const fetchImpl = options.fetch ?? globalFetch;
    if (!fetchImpl) {
      throw new Error(
        "No global fetch available. Use Node 18 or newer, or pass a fetch implementation.",
      );
    }
    const base = (options.baseUrl ?? API_BASE).replace(/\/$/, "");
    const perPage = Math.min(Math.max(options.perPage ?? 200, 1), 200);
    const records: RawDataset[] = [];
    let page = 1;
    let totalPages = 1;

    do {
      const url = `${base}/datasets?page=${page}&per_page=${perPage}`;
      const init: RequestInit = { headers: { accept: "application/json" } };
      if (options.signal) init.signal = options.signal;
      const response = await fetchImpl(url, init);
      if (!response.ok) {
        const body = await response.text().catch(() => "");
        throw new NganoHttpError(response.status, url, body);
      }
      const body = (await response.json()) as {
        data?: RawDataset[];
        meta?: { total_pages?: number };
      };
      records.push(...(body.data ?? []));
      totalPages = body.meta?.total_pages ?? 1;
      page += 1;
    } while (page <= totalPages);

    return new Catalogue(records);
  }

  /**
   * Filters the catalogue.
   *
   * @param filter a {@link Filter} builder or a plain options object
   * @returns matching datasets, in catalogue order
   */
  datasets(filter?: FilterLike): Dataset[] {
    const predicate = compileFilter(toOptions(filter));
    return this.records.filter(predicate);
  }

  /**
   * Looks up one dataset by its catalogue id.
   *
   * @param id ngano catalogue id
   * @returns the dataset, or undefined when the id is unknown
   */
  get(id: string): Dataset | undefined {
    return this.byId.get(id);
  }

  /**
   * Looks up one dataset by its Hugging Face repo id, case-insensitively.
   *
   * @param repo Hugging Face repo id, for example `google/fleurs`
   * @returns the dataset, or undefined when the repo is not catalogued
   */
  getByRepo(repo: string): Dataset | undefined {
    return this.byRepo.get(repo.toLowerCase());
  }

  /**
   * Resolves a country name or code to ISO 3166-1 alpha-2.
   *
   * @param value a country name, alpha-2 or alpha-3 code
   * @returns the alpha-2 code, or null when unrecognised
   */
  resolveCountry(value: string): string | null {
    const country = this.countryIndex.get(fold(value)) ?? this.countryIndex.get(value.toLowerCase());
    return country ? country.iso2 : null;
  }

  /**
   * Summarises the catalogue by country, joining the reference data in
   * `countries.json` with counts derived from the records in hand.
   *
   * @param filter optional filter applied before counting
   * @returns countries with at least one dataset, most datasets first
   */
  countries(filter?: FilterLike): CountrySummary[] {
    const records = this.datasets(filter);
    const counts = new Map<string, { datasets: number; hours: number; languages: Set<string> }>();
    for (const record of records) {
      const codes = record.countryCodes.length
        ? record.countryCodes
        : record.countries.map((name) => this.resolveCountry(name)).filter((c): c is string => !!c);
      for (const code of new Set(codes)) {
        let entry = counts.get(code);
        if (!entry) {
          entry = { datasets: 0, hours: 0, languages: new Set() };
          counts.set(code, entry);
        }
        entry.datasets += 1;
        entry.hours += countableHours(record);
        for (const language of record.languagesClean) entry.languages.add(language);
      }
    }

    const summaries: CountrySummary[] = [];
    for (const country of COUNTRIES as RawCountry[]) {
      const entry = counts.get(country.iso2);
      if (!entry) continue;
      summaries.push({
        name: country.name,
        iso2: country.iso2,
        iso3: country.iso3,
        mapName: country.map_name,
        lat: country.lat,
        lon: country.lon,
        region: country.region,
        slug: country.slug,
        datasets: entry.datasets,
        hours: round(entry.hours),
        languages: [...entry.languages].sort(),
      });
    }
    summaries.sort((a, b) => b.datasets - a.datasets || a.name.localeCompare(b.name));
    return summaries;
  }

  /**
   * Summarises the catalogue by language tag, joining `languages.json` with
   * counts derived from the records in hand. A regional variety such as
   * `eng-NG` is summarised separately from `eng`.
   *
   * @param filter optional filter applied before counting
   * @returns language tags with at least one dataset, most datasets first
   */
  languages(filter?: FilterLike): LanguageSummary[] {
    const records = this.datasets(filter);
    const counts = new Map<string, { datasets: number; hours: number; countries: Set<string> }>();
    for (const record of records) {
      for (const tag of new Set(record.languageTags)) {
        let entry = counts.get(tag);
        if (!entry) {
          entry = { datasets: 0, hours: 0, countries: new Set() };
          counts.set(tag, entry);
        }
        entry.datasets += 1;
        entry.hours += countableHours(record);
        for (const country of record.countries) entry.countries.add(country);
      }
    }

    const summaries: LanguageSummary[] = [];
    for (const [tag, entry] of counts) {
      const reference = this.languageIndex.get(tag.toLowerCase());
      summaries.push({
        tag,
        iso639_3: reference?.iso639_3 ?? primaryCode(tag),
        region: reference?.region ?? null,
        name: reference?.name ?? languageName(tag),
        slug: reference?.slug ?? tag.toLowerCase(),
        datasets: entry.datasets,
        countries: [...entry.countries].sort(),
        hours: round(entry.hours),
      });
    }
    summaries.sort((a, b) => b.datasets - a.datasets || a.name.localeCompare(b.name));
    return summaries;
  }

  /**
   * The distinct bare ISO 639-3 codes in the catalogue, with the regional
   * varieties of a code collapsed onto it, sorted.
   *
   * @param filter optional filter applied before collecting
   * @returns the codes, sorted
   */
  languageCodes(filter?: FilterLike): string[] {
    const codes = new Set<string>();
    for (const record of this.datasets(filter)) {
      for (const code of record.languageCodes) codes.add(code);
    }
    return [...codes].sort();
  }

  /**
   * Resolves a language tag, bare ISO 639-3 code or name to canonical tags.
   *
   * A tag beats a name, a bare code that only exists as regional varieties
   * resolves to those varieties, and anything unrecognised resolves to nothing
   * rather than to a guess.
   *
   * @param value a tag, a code or any catalogue spelling of a name
   * @param options whether a bare code widens to its regional varieties
   * @returns canonical tags, possibly empty
   *
   * @example
   * ```ts
   * catalogue.resolveLanguage("isiZulu");  // ["zul"]
   * catalogue.resolveLanguage("ENG-ng");   // ["eng-NG"]
   * ```
   */
  resolveLanguage(value: string, options: ResolveOptions = {}): string[] {
    return resolveLanguageTags(value, options);
  }

  /**
   * Aggregate counts. Hours exclude self-reported figures of 20,000 hours or
   * more, which the catalogue flags as unverified.
   *
   * @param filter optional filter applied before counting
   * @returns the aggregates
   */
  stats(filter?: FilterLike): Stats {
    const records = this.datasets(filter);
    const tally = (
      bucket: Record<string, number>,
      value: string | null | undefined,
    ): void => {
      const key = value && value.length > 0 ? value : "Unstated";
      bucket[key] = (bucket[key] ?? 0) + 1;
    };

    const stats: Stats = {
      datasets: records.length,
      hours: 0,
      unverifiedExcluded: 0,
      languages: 0,
      languageCodes: 0,
      countries: 0,
      hfRepos: 0,
      byTask: {},
      byAccess: {},
      byCommercial: {},
      byLicenceClass: {},
      byQuality: {},
      byLabelled: {},
      byRegion: {},
    };

    const languages = new Set<string>();
    const languageCodes = new Set<string>();
    const countries = new Set<string>();
    for (const record of records) {
      stats.hours += countableHours(record);
      if (record.unverifiedSize) stats.unverifiedExcluded += 1;
      if (record.hfRepo) stats.hfRepos += 1;
      for (const tag of record.languageTags) languages.add(tag);
      for (const code of record.languageCodes) languageCodes.add(code);
      for (const code of record.countryCodes) countries.add(code);
      tally(stats.byTask, record.task);
      tally(stats.byAccess, record.access);
      tally(stats.byCommercial, record.commercial);
      tally(stats.byLicenceClass, record.licenceClass);
      tally(stats.byQuality, record.quality);
      tally(stats.byLabelled, record.labelled);
      for (const region of new Set(record.regions)) tally(stats.byRegion, region);
    }
    stats.hours = round(stats.hours);
    stats.languages = languages.size;
    stats.languageCodes = languageCodes.size;
    stats.countries = countries.size;
    return stats;
  }

  /**
   * Free-text search over names, ids, notes, languages and countries. Name and
   * id hits rank above notes hits.
   *
   * @param query the search text
   * @param filter optional filter applied before searching
   * @returns matching datasets, best first
   */
  search(query: string, filter?: FilterLike): Dataset[] {
    const needle = query.trim().toLowerCase();
    if (needle === "") return this.datasets(filter);
    const folded = fold(query);
    const scored: Array<{ dataset: Dataset; score: number }> = [];
    for (const dataset of this.datasets(filter)) {
      const name = dataset.name.toLowerCase();
      let score = 0;
      if (name === needle || dataset.id === needle) score = 100;
      else if (name.startsWith(needle)) score = 80;
      else if (name.includes(needle) || fold(dataset.name).includes(folded)) score = 60;
      else if (dataset.id.includes(needle)) score = 55;
      else if (dataset.languageTags.some((tag) => tag.toLowerCase() === needle)) score = 45;
      else if (dataset.languages.some((language) => fold(language).includes(folded))) score = 40;
      else if (dataset.languagesClean.some((language) => fold(language).includes(folded))) score = 38;
      else if (dataset.countries.some((country) => fold(country).includes(folded))) score = 30;
      else if ((dataset.hfRepo ?? "").toLowerCase().includes(needle)) score = 25;
      else if ((dataset.notes ?? "").toLowerCase().includes(needle)) score = 20;
      else if ((dataset.domain ?? "").toLowerCase().includes(needle)) score = 15;
      if (score > 0) scored.push({ dataset, score });
    }
    scored.sort(
      (a, b) =>
        b.score - a.score ||
        countableHours(b.dataset) - countableHours(a.dataset) ||
        a.dataset.name.localeCompare(b.dataset.name),
    );
    return scored.map((entry) => entry.dataset);
  }
}

/**
 * Rounds to one decimal place, which is as much precision as published hours
 * figures ever carry.
 *
 * @param value a number
 * @returns the rounded number
 */
function round(value: number): number {
  return Math.round(value * 10) / 10;
}

/** A catalogue over the bundled snapshot, created once and shared. */
let defaultCatalogue: Catalogue | null = null;

/**
 * The shared offline catalogue. Building it parses the bundled snapshot, so it
 * is done once and reused.
 *
 * @returns the shared catalogue
 */
export function getDefaultCatalogue(): Catalogue {
  defaultCatalogue ??= new Catalogue();
  return defaultCatalogue;
}

File: packages/rust/src/catalogue.rs
//! The dataset catalogue: bundled offline snapshot, or live from the HTTP API.

use std::collections::{BTreeMap, BTreeSet, HashMap};
use std::sync::Arc;

use serde::Deserialize;

use crate::error::{NganoError, Result};
use crate::filter::Filter;
use crate::http::HttpClient;
use crate::language::{LanguageCode, LanguageRegistry};
use crate::model::{Country, Dataset, Language, Stats};

/// The public ngano API base URL.
pub const DEFAULT_API_BASE: &str = "https://ngano.dev/api/v1";

/// The catalogue snapshot compiled into this crate.
const BUNDLED_CATALOGUE: &str = include_str!(concat!(env!("OUT_DIR"), "/catalogue.json"));
/// The country list compiled into this crate.
const BUNDLED_COUNTRIES: &str = include_str!(concat!(env!("OUT_DIR"), "/countries.json"));
/// The language list compiled into this crate.
const BUNDLED_LANGUAGES: &str = include_str!(concat!(env!("OUT_DIR"), "/languages.json"));

/// Envelope returned by `GET /datasets`.
#[derive(Debug, Deserialize)]
struct Page {
    data: Vec<Dataset>,
    #[serde(default)]
    meta: PageMeta,
}

/// Pagination block of a `GET /datasets` response.
#[derive(Debug, Default, Deserialize)]
struct PageMeta {
    #[serde(default)]
    total_pages: Option<u32>,
}

/// An in-memory catalogue of datasets, countries and languages.
///
/// Cloning is cheap: the records sit behind an [`Arc`].
///
/// ```
/// use ngano::{Catalogue, Filter};
///
/// let cat = Catalogue::bundled()?;
/// assert!(cat.len() > 100);
/// let shona = cat.datasets(&Filter::new().language("sna"));
/// assert!(!shona.is_empty());
/// # Ok::<(), ngano::NganoError>(())
/// ```
#[derive(Debug, Clone)]
pub struct Catalogue {
    inner: Arc<Inner>,
}

/// The owned contents of a catalogue.
#[derive(Debug)]
pub(crate) struct Inner {
    datasets: Vec<Dataset>,
    countries: Vec<Country>,
    languages: Vec<Language>,
    by_id: HashMap<String, usize>,
    by_repo: HashMap<String, usize>,
    country_by_name: HashMap<String, String>,
    language_by_tag: HashMap<String, usize>,
}

impl Catalogue {
    /// Load the snapshot compiled into the binary. Never touches the network.
    ///
    /// ```
    /// let cat = ngano::Catalogue::bundled()?;
    /// assert!(!cat.countries().is_empty());
    /// # Ok::<(), ngano::NganoError>(())
    /// ```
    pub fn bundled() -> Result<Self> {
        let datasets: Vec<Dataset> = serde_json::from_str(BUNDLED_CATALOGUE)?;
        let countries: Vec<Country> = serde_json::from_str(BUNDLED_COUNTRIES)?;
        let languages: Vec<Language> = serde_json::from_str(BUNDLED_LANGUAGES)?;
        Ok(Self::from_parts(datasets, countries, languages))
    }

    /// Build a catalogue from records you already hold.
    pub fn from_parts(
        datasets: Vec<Dataset>,
        countries: Vec<Country>,
        languages: Vec<Language>,
    ) -> Self {
        let by_id = datasets
            .iter()
            .enumerate()
            .map(|(i, d)| (d.id.to_lowercase(), i))
            .collect();
        let by_repo = datasets
            .iter()
            .enumerate()
            .filter_map(|(i, d)| Some((d.hf_repo.as_ref()?.to_lowercase(), i)))
            .collect();
        let mut country_by_name = HashMap::with_capacity(countries.len() * 3);
        for c in &countries {
            country_by_name.insert(c.name.to_lowercase(), c.iso2.clone());
            if let Some(map_name) = &c.map_name {
                country_by_name.insert(map_name.to_lowercase(), c.iso2.clone());
            }
            country_by_name.insert(c.iso3.to_lowercase(), c.iso2.clone());
        }
        let language_by_tag = languages
            .iter()
            .enumerate()
            .map(|(i, l)| (l.tag.to_lowercase(), i))
            .collect();
        Self {
            inner: Arc::new(Inner {
                datasets,
                countries,
                languages,
                by_id,
                by_repo,
                country_by_name,
                language_by_tag,
            }),
        }
    }

    /// Fetch the live catalogue from `https://ngano.dev/api/v1`.
    ///
    /// ```no_run
    /// # async fn run() -> Result<(), ngano::NganoError> {
    /// let cat = ngano::Catalogue::from_api().await?;
    /// println!("{} datasets", cat.len());
    /// # Ok(()) }
    /// ```
    pub async fn from_api() -> Result<Self> {
        Self::from_api_base(DEFAULT_API_BASE).await
    }

    /// Fetch the live catalogue from a specific API base URL, for a mirror or a
    /// local development server.
    pub async fn from_api_base(base: &str) -> Result<Self> {
        let base = base.trim_end_matches('/');
        let http = HttpClient::new("ngano api")?;

        let mut datasets = Vec::new();
        let mut page = 1u32;
        loop {
            let url = format!("{base}/datasets?page={page}&per_page=200");
            let body: Page = http.get_json(&url, None).await?;
            let empty = body.data.is_empty();
            datasets.extend(body.data);
            let total_pages = body.meta.total_pages.unwrap_or(page);
            if empty || page >= total_pages {
                break;
            }
            page += 1;
        }

        let countries: Vec<Country> = http.get_json(&format!("{base}/countries"), None).await?;
        let languages: Vec<Language> = http.get_json(&format!("{base}/languages"), None).await?;
        Ok(Self::from_parts(datasets, countries, languages))
    }

    /// Blocking counterpart of [`Catalogue::from_api`].
    ///
    /// Available with the `blocking` feature. It starts a private current-thread
    /// runtime, so it must not be called from inside an async runtime.
    #[cfg(feature = "blocking")]
    #[cfg_attr(docsrs, doc(cfg(feature = "blocking")))]
    pub fn from_api_blocking() -> Result<Self> {
        crate::blocking::block_on(Self::from_api())
    }

    /// Blocking counterpart of [`Catalogue::from_api_base`].
    #[cfg(feature = "blocking")]
    #[cfg_attr(docsrs, doc(cfg(feature = "blocking")))]
    pub fn from_api_base_blocking(base: &str) -> Result<Self> {
        crate::blocking::block_on(Self::from_api_base(base))
    }

    /// Every dataset record, in catalogue order.
    pub fn all(&self) -> &[Dataset] {
        &self.inner.datasets
    }

    /// Number of dataset records.
    pub fn len(&self) -> usize {
        self.inner.datasets.len()
    }

    /// True when the catalogue holds no dataset records.
    pub fn is_empty(&self) -> bool {
        self.inner.datasets.is_empty()
    }

    /// Every country in the coverage map.
    pub fn countries(&self) -> &[Country] {
        &self.inner.countries
    }

    /// Every language in the coverage map, one per BCP 47 tag.
    pub fn languages(&self) -> &[Language] {
        &self.inner.languages
    }

    /// The ISO 639-3 registry compiled into this crate: every canonical tag,
    /// with its bare code, its name and every catalogue spelling of it.
    ///
    /// This is the registry itself, so it lists tags the catalogue may not have
    /// a dataset for, whereas [`Catalogue::languages`] lists coverage.
    ///
    /// ```
    /// let cat = ngano::Catalogue::bundled()?;
    /// let sna = cat
    ///     .language_codes()
    ///     .iter()
    ///     .find(|c| c.tag == "sna")
    ///     .expect("Shona is in the registry");
    /// assert_eq!(sna.name, "Shona");
    /// assert_eq!(sna.iso639_3, "sna");
    /// # Ok::<(), ngano::NganoError>(())
    /// ```
    pub fn language_codes(&self) -> &[LanguageCode] {
        self.registry().codes()
    }

    /// The language registry behind [`Catalogue::language_codes`].
    pub fn registry(&self) -> &'static LanguageRegistry {
        LanguageRegistry::shared()
    }

    /// Resolve a code, a tag or a name to canonical BCP 47 tags.
    ///
    /// The rules are [`LanguageRegistry::resolve`], without widening: a tag
    /// first, then a bare code that exists only as regional varieties, then a
    /// name from the registry aliases. Anything unrecognised gives an empty
    /// vector rather than a guess.
    ///
    /// ```
    /// let cat = ngano::Catalogue::bundled()?;
    /// assert_eq!(cat.resolve_language("isiZulu"), vec!["zul".to_string()]);
    /// assert_eq!(cat.resolve_language("ENG-ng"), vec!["eng-NG".to_string()]);
    /// assert!(cat.resolve_language("Klingon").is_empty());
    /// # Ok::<(), ngano::NganoError>(())
    /// ```
    pub fn resolve_language(&self, value: &str) -> Vec<String> {
        self.registry().resolve(value, false)
    }

    /// The coverage record for any accepted spelling of a language: a tag, a
    /// bare code or a name. Never widens a bare code to its varieties.
    ///
    /// ```
    /// let cat = ngano::Catalogue::bundled()?;
    /// assert_eq!(cat.language("Shona").map(|l| l.tag.as_str()), Some("sna"));
    /// assert!(cat.language("nothing-at-all").is_none());
    /// # Ok::<(), ngano::NganoError>(())
    /// ```
    pub fn language(&self, value: &str) -> Option<&Language> {
        self.resolve_language(value).into_iter().find_map(|tag| {
            self.inner
                .language_by_tag
                .get(&tag.to_lowercase())
                .map(|i| &self.inner.languages[*i])
        })
    }

    /// Datasets matching a filter, in catalogue order.
    ///
    /// ```
    /// use ngano::{Catalogue, Filter};
    ///
    /// let cat = Catalogue::bundled()?;
    /// let hits = cat.datasets(&Filter::new().country("ZW").hf_only(true));
    /// assert!(hits.iter().all(|d| d.is_streamable()));
    /// # Ok::<(), ngano::NganoError>(())
    /// ```
    pub fn datasets(&self, filter: &Filter) -> Vec<&Dataset> {
        self.inner
            .datasets
            .iter()
            .filter(|d| filter.matches(d))
            .collect()
    }

    /// One dataset by its catalogue id, case-insensitively.
    ///
    /// ```
    /// let cat = ngano::Catalogue::bundled()?;
    /// let first = cat.all().first().unwrap().id.clone();
    /// assert!(cat.get(&first).is_some());
    /// assert!(cat.get("no-such-dataset").is_none());
    /// # Ok::<(), ngano::NganoError>(())
    /// ```
    pub fn get(&self, id: &str) -> Option<&Dataset> {
        self.inner
            .by_id
            .get(&id.to_lowercase())
            .map(|i| &self.inner.datasets[*i])
    }

    /// One dataset by its Hugging Face repo id, case-insensitively.
    ///
    /// ```
    /// let cat = ngano::Catalogue::bundled()?;
    /// let repo = cat
    ///     .all()
    ///     .iter()
    ///     .find_map(|d| d.hf_repo.clone())
    ///     .expect("the catalogue lists hugging face repos");
    /// assert!(cat.by_hf_repo(&repo).is_some());
    /// assert!(cat.by_hf_repo("nobody/nothing").is_none());
    /// # Ok::<(), ngano::NganoError>(())
    /// ```
    pub fn by_hf_repo(&self, repo: &str) -> Option<&Dataset> {
        self.inner
            .by_repo
            .get(&repo.trim().to_lowercase())
            .map(|i| &self.inner.datasets[*i])
    }

    /// One dataset by id, as an error when it is missing.
    pub fn require(&self, id: &str) -> Result<&Dataset> {
        self.get(id)
            .ok_or_else(|| NganoError::NotFound(format!("dataset {id}")))
    }

    /// Free-text search over names, ids, languages, countries and notes.
    ///
    /// ```
    /// let cat = ngano::Catalogue::bundled()?;
    /// let hits = cat.search("parliament");
    /// assert!(hits.iter().all(|d| !d.id.is_empty()));
    /// # Ok::<(), ngano::NganoError>(())
    /// ```
    pub fn search(&self, query: &str) -> Vec<&Dataset> {
        self.datasets(&Filter::new().q(query))
    }

    /// Aggregates over the whole catalogue. Hours exclude unverified figures.
    ///
    /// ```
    /// let cat = ngano::Catalogue::bundled()?;
    /// let s = cat.stats();
    /// assert_eq!(s.datasets, cat.len());
    /// assert!(s.hours > 0.0);
    /// # Ok::<(), ngano::NganoError>(())
    /// ```
    pub fn stats(&self) -> Stats {
        self.stats_for(self.all())
    }

    /// Aggregates over an arbitrary selection, such as the result of
    /// [`Catalogue::datasets`].
    pub fn stats_for<'a, I>(&self, datasets: I) -> Stats
    where
        I: IntoIterator<Item = &'a Dataset>,
    {
        let mut s = Stats {
            datasets: 0,
            hours: 0.0,
            unverified_datasets: 0,
            streamable_datasets: 0,
            languages: 0,
            language_codes: 0,
            countries: 0,
            by_task: BTreeMap::new(),
            by_access: BTreeMap::new(),
            by_licence_class: BTreeMap::new(),
            by_commercial: BTreeMap::new(),
            by_region: BTreeMap::new(),
            hours_by_region: BTreeMap::new(),
        };
        let mut languages = BTreeSet::new();
        let mut language_codes = BTreeSet::new();
        let mut countries = BTreeSet::new();

        for d in datasets {
            s.datasets += 1;
            if d.unverified_size {
                s.unverified_datasets += 1;
            }
            if d.is_streamable() {
                s.streamable_datasets += 1;
            }
            let hours = d.countable_hours().unwrap_or(0.0);
            s.hours += hours;
            for tag in &d.language_tags {
                languages.insert(tag.clone());
            }
            for code in &d.language_codes {
                language_codes.insert(code.clone());
            }
            for c in &d.country_codes {
                countries.insert(c.clone());
            }
            *s.by_task.entry(d.task.clone()).or_default() += 1;
            *s.by_access.entry(d.access.clone()).or_default() += 1;
            *s.by_licence_class
                .entry(d.licence_class.clone())
                .or_default() += 1;
            *s.by_commercial.entry(d.commercial.clone()).or_default() += 1;
            for r in &d.regions {
                *s.by_region.entry(r.clone()).or_default() += 1;
                *s.hours_by_region.entry(r.clone()).or_default() += hours;
            }
        }
        s.languages = languages.len();
        s.language_codes = language_codes.len();
        s.countries = countries.len();
        s
    }

    /// Resolve a country name or code to an ISO 3166-1 alpha-2 code.
    ///
    /// ```
    /// let cat = ngano::Catalogue::bundled()?;
    /// assert_eq!(cat.country_code("Zimbabwe").as_deref(), Some("ZW"));
    /// assert_eq!(cat.country_code("zw").as_deref(), Some("ZW"));
    /// assert_eq!(cat.country_code("Atlantis"), None);
    /// # Ok::<(), ngano::NganoError>(())
    /// ```
    pub fn country_code(&self, value: &str) -> Option<String> {
        let trimmed = value.trim();
        if trimmed.len() == 2 {
            let upper = trimmed.to_uppercase();
            if self.inner.countries.iter().any(|c| c.iso2 == upper) {
                return Some(upper);
            }
        }
        self.inner
            .country_by_name
            .get(&trimmed.to_lowercase())
            .cloned()
    }

    /// Internal handle on the name-to-code table, shared with row mapping.
    pub(crate) fn country_index(&self) -> Arc<Inner> {
        Arc::clone(&self.inner)
    }
}

impl Inner {
    /// Resolve a country label seen in a data row to an alpha-2 code.
    pub(crate) fn resolve_country(&self, value: &str) -> Option<String> {
        let trimmed = value.trim();
        if trimmed.is_empty() {
            return None;
        }
        if trimmed.len() == 2 {
            let upper = trimmed.to_uppercase();
            if self.countries.iter().any(|c| c.iso2 == upper) {
                return Some(upper);
            }
        }
        self.country_by_name.get(&trimmed.to_lowercase()).cloned()
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn bundled_catalogue_parses() {
        let cat = Catalogue::bundled().expect("bundled catalogue parses");
        assert!(
            cat.len() >= 600,
            "expected the full catalogue, got {}",
            cat.len()
        );
        assert_eq!(cat.countries().len(), 58);
        assert!(cat.languages().len() >= 300);
        assert!(cat
            .languages()
            .iter()
            .all(|l| l.slug == l.tag.to_lowercase()));
    }

    #[test]
    fn languages_resolve_through_the_registry() {
        let cat = Catalogue::bundled().unwrap();
        assert_eq!(cat.resolve_language("isiZulu"), vec!["zul".to_string()]);
        assert_eq!(cat.resolve_language("eng-ng"), vec!["eng-NG".to_string()]);
        assert!(cat.resolve_language("").is_empty());
        assert_eq!(cat.language("sna").map(|l| l.name.as_str()), Some("Shona"));
        assert!(cat.language_codes().len() >= 300);
    }

    #[test]
    fn unverified_hours_are_excluded_from_totals() {
        let cat = Catalogue::bundled().unwrap();
        let naive: f64 = cat.all().iter().filter_map(|d| d.hours_num).sum();
        let counted = cat.stats().hours;
        assert!(counted < naive, "unverified hours must be excluded");
        for d in cat.all().iter().filter(|d| d.unverified_size) {
            assert_eq!(d.countable_hours(), None);
        }
    }
}
